### PR TITLE
[cryptotest] Test harness for SHA2/3 and SHAKE

### PR DESCRIPTION
**Pull request by RyanTorok**
This PR adds the test harness for the SHA2/3 and SHAKE hash functions, using the NIST test vectors added in #21261. The test implementation contains a Rust host-side test driver, a device-side firmware handler, and a uJSON API allowing the two to communicate.

The SHA-2 tests test both the oneshot API and the stepwise APIs (init, update, final). At present, SHA-3 and SHAKE only support the oneshot API, so it is the only one tested.

The hash tests take about an hour to run in total, so this will certainly need to go in the nightly CI run eventually.

~Dependent on #21261~

### CODE REVIEW

**Comment by jadephilipoom** [comment]
> The hash tests take about an hour to run in total, so this will certainly need to go in the nightly CI run eventually.

As written, does this PR add the tests to CI? If so we may want to move them to nightly before merging it.

File: sw/device/tests/crypto/cryptotest/firmware/hash.c [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+      break;
+    case kCryptotestHashAlgorithmSha3_512:
+      mode = kOtcryptoHashModeSha3_512;
+      digest.len = kOtcryptoHashDigestWordsSha3_512;
+      hash_oneshot = otcrypto_hash;
+      break;
+    case kCryptotestHashAlgorithmShake128:
+      mode = kOtcryptoHashXofModeShake128;
+      // length in words = ceil(length in bytes / word size)
+      digest.len = (uj_shake_digest_length.length + 3) / sizeof(uint32_t);
```
Comment (milesdai) [inline sw/device/tests/crypto/cryptotest/firmware/hash.c:98]: There exists a [`ceil_div`](https://github.com/lowRISC/opentitan/blob/6f251215344949e9dcc596c67860defcba856d34/sw/device/lib/base/math.h#L60) implementation in math.h.

File: sw/device/tests/crypto/cryptotest/firmware/hash.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#include "sw/device/lib/testing/test_framework/ujson_ottf.h"
+#include "sw/device/lib/ujson/ujson.h"
+#include "sw/device/tests/crypto/cryptotest/json/hash_commands.h"
+
+const int MaxMessageBytes = 17068;
+
+const int MaxDigestBytes = 256;
+const int MaxDigestWords = MaxDigestBytes / sizeof(uint32_t);
+
+const unsigned int kOtcryptoHashDigestWordsSha256 = 8;
```
Comment (milesdai) [inline sw/device/tests/crypto/cryptotest/firmware/hash.c:22]: These constants are defined in [sha256.h](https://github.com/lowRISC/opentitan/blob/936d2a10e159e6096adec628ef69a67a82ec7ac5/sw/device/lib/crypto/impl/sha2/sha256.h#L52)

File: sw/device/tests/crypto/cryptotest/firmware/hash.c [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+#include "sw/device/lib/crypto/include/datatypes.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/ujson_ottf.h"
+#include "sw/device/lib/ujson/ujson.h"
+#include "sw/device/tests/crypto/cryptotest/json/hash_commands.h"
+
+const int MaxMessageBytes = 17068;
+
+const int MaxDigestBytes = 256;
+const int MaxDigestWords = MaxDigestBytes / sizeof(uint32_t);
```
Comment (milesdai) [inline sw/device/tests/crypto/cryptotest/firmware/hash.c:20]: Add a `k` to the start of these variable names and define in an enum:

```C
enum {
  kMaxMessageBytes = 17068,
  ...
};

**Review by milesdai — APPROVED** [review]
LGTM! Just a few nits.

File: sw/device/tests/crypto/cryptotest/firmware/hash.c (hunk 2 above)
Comment (RyanTorok) [reply sw/device/tests/crypto/cryptotest/firmware/hash.c:22]: Strangely, they only are for `SHA-256` and `SHA-512`. I still had to hard-code the others.

**Review by jadephilipoom — CHANGES_REQUESTED** [review]
Mostly looks good, but I think as currently written this could cause buffer overflow if new test vectors are added that have longer messages/digests.

File: sw/device/tests/crypto/cryptotest/firmware/hash.c [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+#include "sw/device/lib/crypto/impl/keyblob.h"
+#include "sw/device/lib/crypto/include/datatypes.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/ujson_ottf.h"
+#include "sw/device/lib/ujson/ujson.h"
+#include "sw/device/tests/crypto/cryptotest/json/hash_commands.h"
+
+const int MaxMessageBytes = 17068;
+
+const int MaxDigestBytes = 256;
```
[the remark this answers is not in the document]
Comment (RyanTorok) [reply sw/device/tests/crypto/cryptotest/firmware/hash.c]: Once again I've outed myself as a Rust programmer. I constantly think it's black magic to see a C program stack-allocate an array with the size determined at runtime.

**Review by jadephilipoom — APPROVED** [review]

**Pull request merged**